src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1563,16 +1564,29 @@ internal void RemoveProjectFromMaps(CodeAnalysis.Project project)
 
     internal async Task RefreshProjectExistsUIContextForLanguageAsync(string language, CancellationToken cancellationToken)
     {
-        await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
+        UIContext? uiContext;
 
-        var uiContext = _languageToProjectExistsUIContext.GetOrAdd(
-            language,
-            language => Services.GetLanguageServices(language).GetService<IProjectExistsUIContextProviderLanguageService>()?.GetUIContext());
+        using (_gate.DisposableWait(cancellationToken))
+        {
+            (uiContext, var isContextActive) = _languageToProjectExistsUIContext.GetOrAdd(
+                language,
+                language => (Services.GetLanguageServices(language).GetService<IProjectExistsUIContextProviderLanguageService>()?.GetUIContext(), false));
```

Comment:
Is fetching the UI context itself free-threaded? I feel like when this was originally written that wasn't the case.

Reply:
(and if it's not, the fix is simple to just defer that bit to the UI thread...)

Reply:
And also frankly now since we're avoiding unnecessary updates, you could just remove the UIContext from the map entirely, since we're only going to fetch it 'rarely'.

Reply:
I thought so because of this comment:

[https://devdiv.visualstudio.com/DevDiv/_git/VS?path=/src/env/shell/PackageFramework/Framework/UIContextImpl.cs&version=GBmain&_a=contents&line=63&lineStyle=plain&lineEnd=70&lineStartColumn=1&lineEndColumn=1](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=/src/env/shell/PackageFramework/Framework/UIContextImpl.cs&version=GBmain&_a=contents&line=63&lineStyle=plain&lineEnd=70&lineStartColumn=1&lineEndColumn=1)

Reply:
Although I do believe this is safe because of the comment linked to above, I do prefer your idea of just retrieving the UIContext while on the main thread, instead of caching it. It simplifies the data structures, and the removes the dependency that creating the ui context is free threaded.